File: go/tree.go
package gencmu

import "strings"

// The tree (engine §12) and emission (engine §11), both walks of a chosen
// derivation. Derivations nest as deep as a text is long, so the walks keep
// their own stacks.

// flattenKids lists a close's children in order.
func flattenKids(kids *dn) []*dn {
	var out []*dn
	for p := kids; p != nil; p = p.a {
		out = append(out, p.b)
	}
	for i, j := 0, len(out)-1; i < j; i, j = i+1, j-1 {
		out[i], out[j] = out[j], out[i]
	}
	return out
}

// emptySource is the source of an empty span at token position p: the end
// of the token before it, or the start of the first token.
func (run *stageRun) emptySource(p int) [2]int {
	switch {
	case p > 0 && p <= len(run.toks):
		e := run.toks[p-1].Source[1]
		return [2]int{e, e}
	case len(run.toks) > 0:
		s := run.toks[0].Source[0]
		return [2]int{s, s}
	}
	return [2]int{0, 0}
}

func (run *stageRun) spanSource(a, b int) [2]int {
	if b <= a {
		return run.emptySource(a)
	}
	return [2]int{run.toks[a].Source[0], run.toks[b-1].Source[1]}
}

type pendingKid struct {
	n      *dn
	splice bool // the prefix of a trailing repetition
}

// treeFrame is a rule node being built; pending holds its remaining
// children in reverse, the next one last.
type treeFrame struct {
	node    *Node
	pending []pendingKid
}

// pushKids pushes a close's children onto a pending stack, the first last,
// marking the first as spliced when the close is a trailing repetition's.
func pushKids(pending []pendingKid, n *dn, splice bool) []pendingKid {
	kids := flattenKids(n.a)
	for i := len(kids) - 1; i >= 0; i-- {
		pending = append(pending, pendingKid{n: kids[i], splice: splice && i == 0 && n.prod.repeatPrefix})
	}
	return pending
}

// buildTree turns a derivation of a constituent into the result's tree.
func (run *stageRun) buildTree(rec *recognizer, d *dn) *Node {
	base := rec.base
	g := rec.g
	newRule := func(n *dn) *treeFrame {
		a, b := base+int(n.start), base+int(n.end)
		node := &Node{Kind: KindRule, Rule: n.prod.ruleName, Span: [2]int{a, b}, Source: run.spanSource(a, b), Tags: n.tags.toMap(), Children: []*Node{}}
		return &treeFrame{node: node, pending: pushKids(nil, n, true)}
	}
	root := newRule(d)
	stack := []*treeFrame{root}
	for len(stack) > 0 {
		f := stack[len(stack)-1]
		if len(f.pending) == 0 {
			stack = stack[:len(stack)-1]
			if len(stack) > 0 {
				parent := stack[len(stack)-1]
				parent.node.Children = append(parent.node.Children, f.node)
			}
			continue
		}
		k := f.pending[len(f.pending)-1]
		f.pending = f.pending[:len(f.pending)-1]
		n := k.n
		switch {
		case n.kind == dRead:
			i := base + int(n.tok)
			f.node.Children = append(f.node.Children, &Node{Kind: KindToken, Terminal: g.terminals[n.term], Token: i, Span: [2]int{i, i + 1}, Source: run.toks[i].Source})
		case n.prod.helper && n.a == nil && n.prod.elided != "":
			p := base + int(n.start)
			f.node.Children = append(f.node.Children, &Node{Kind: KindElided, Terminal: n.prod.elided, Span: [2]int{p, p}, Source: run.emptySource(p)})
		case n.prod.helper || k.splice:
			f.pending = pushKids(f.pending, n, k.splice)
		default:
			stack = append(stack, newRule(n))
		}
	}
	return root.node
}

// elidedNodes lists a tree's elided nodes in text order.
func elidedNodes(root *Node) []*Node {
	var out []*Node
	stack := []*Node{root}
	for len(stack) > 0 {
		n := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		if n.Kind == KindElided {
			out = append(out, n)
		}
		for i := len(n.Children) - 1; i >= 0; i-- {
			stack = append(stack, n.Children[i])
		}
	}
	return out
}

// ---- emission

type emitTask struct {
	walk *dn
	emit *dn
	tags *tagset
	tok  *Token
}

func (run *stageRun) kidSpan(rec *recognizer, k *dn) (int, int, *tagset) {
	if k.kind == dRead {
		i := rec.base + int(k.tok)
		return i, i + 1, run.tagsets[i]
	}
	return rec.base + int(k.start), rec.base + int(k.end), k.tags
}

// emit walks the chosen derivation from the left and returns the tokens of
// the next stage.
func (run *stageRun) emit(rec *recognizer, d *dn) []Token {
	out := []Token{}
	stack := []emitTask{{walk: d}}
	for len(stack) > 0 {
		t := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		switch {
		case t.tok != nil:
			out = append(out, *t.tok)
		case t.emit != nil:
			out = append(out, run.emitted(rec, t.emit, t.tags))
		case t.walk != nil:
			n := t.walk
			if n.kind == dRead {
				continue
			}
			plan := run.plan(rec, n)
			for i := len(plan) - 1; i >= 0; i-- {
				stack = append(stack, plan[i])
			}
		}
	}
	return out
}

// plan is what one constituent's emission clause does, in order.
func (run *stageRun) plan(rec *recognizer, n *dn) []emitTask {
	p := n.prod
	kids := flattenKids(n.a)
	if p.emit == nil {
		plan := make([]emitTask, len(kids))
		for i, k := range kids {
			plan[i] = emitTask{walk: k}
		}
		return plan
	}
	if p.eraseAll {
		return nil
	}
	start, end := rec.base+int(n.start), rec.base+int(n.end)
	ev := run.evaluator(rec.g, func(name string) (spanVal, bool) {
		if name == "" {
			return spanVal{a: start, b: end, whole: true, tags: n.tags}, true
		}
		for i, c := range p.capName {
			if c == name {
				a, b, tags := run.kidSpan(rec, kids[i])
				return spanVal{a: a, b: b, whole: true, tags: tags}, true
			}
		}
		return spanVal{}, false
	})
	// An item's tag term that gives no tags is an error of the grammar: no
	// terminal could read the token (§11).
	itemTags := func(it *domEmitItem, own *tagset) *tagset {
		if it.Tags == nil {
			return own
		}
		tags := ev.tagsOf(it.Tags)
		if len(tags.names) == 0 {
			panic(&parseFailure{message: p.ruleName + " emits a token with no tags", token: start, tokenEnd: end, hasToken: true, rule: p.ruleName})
		}
		return tags
	}
	items := p.emit.Items
	if p.emit.whole() {
		var plan []emitTask
		for _, it := range items {
			plan = append(plan, emitTask{emit: n, tags: itemTags(it, n.tags)})
		}
		return plan
	}
	insert := func(tag string, index int) emitTask {
		src := run.spanSource(start, end)
		if index > start {
			e := run.toks[index-1].Source[1]
			src = [2]int{e, e}
		} else {
			src = [2]int{src[0], src[0]}
		}
		tok := &Token{Text: "", Tags: map[string]bool{tag: true}, Span: [2]int{index, index}, Source: src, InsertedBy: p.ruleName}
		if ph, ok := phonemeTag(tag); ok {
			tok.Phonemes = ph
		}
		return emitTask{tok: tok}
	}
	// An inserted tag goes before the token of the first capture listed
	// after it, or, with none listed after it, after the last child (§11).
	named := map[string]*domEmitItem{}
	before := map[string][]string{}
	var pendingTags []string
	for _, it := range items {
		switch {
		case it.IsInsert:
			pendingTags = append(pendingTags, it.Insert)
		default:
			named[it.Capture] = it
			before[it.Capture] = append(before[it.Capture], pendingTags...)
			pendingTags = nil
		}
	}
	var plan []emitTask
	for i, k := range kids {
		name := p.capName[i]
		it := named[name]
		if name == "" || it == nil {
			plan = append(plan, emitTask{walk: k})
			continue
		}
		a, _, own := run.kidSpan(rec, k)
		for _, tag := range before[name] {
			plan = append(plan, insert(tag, a))
		}
		// An erased capture is neither emitted nor walked; the tags
		// inserted before it stand where its token would have.
		if !it.Erase {
			plan = append(plan, emitTask{emit: k, tags: itemTags(it, own)})
		}
	}
	for _, tag := range pendingTags {
		plan = append(plan, insert(tag, end))
	}
	return plan
}

// emitted is the token a constituent emits, with the given tags.
func (run *stageRun) emitted(rec *recognizer, n *dn, tags *tagset) Token {
	a, b, _ := run.kidSpan(rec, n)
	src := run.spanSource(a, b)
	tok := Token{Text: string(run.ps.text[src[0]:src[1]]), Tags: tags.toMap(), Span: [2]int{a, b}, Source: src}
	var strong []string
	for i, name := range tags.names {
		if ph, ok := phonemeTag(name); ok && tags.strong[i] {
			strong = append(strong, ph)
		}
	}
	if len(strong) > 1 {
		panic(&parseFailure{message: "an emitted token has two strong phoneme tags", token: a, tokenEnd: b, hasToken: true})
	}
	if len(strong) == 1 {
		tok.Phonemes = strong[0]
		return tok
	}
	// The phonemes of the tokens it was emitted from, omitting every token
	// inside an erased constituent.
	var sb strings.Builder
	stack := []*dn{n}
	for len(stack) > 0 {
		x := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		switch x.kind {
		case dRead:
			sb.WriteString(run.toks[rec.base+int(x.tok)].Phonemes)
		case dClose:
			if x.prod.eraseAll || x.a == nil {
				continue
			}
			if x.prod.erased != nil {
				kids := flattenKids(x.a)
				for i := len(kids) - 1; i >= 0; i-- {
					if !x.prod.erased[i] {
						stack = append(stack, kids[i])
					}
				}
				continue
			}
			stack = append(stack, x.a)
		case dPart:
			stack = append(stack, x.b)
			if x.a != nil {
				stack = append(stack, x.a)
			}
		}
	}
	tok.Phonemes = strings.Trim(sb.String(), " ")
	return tok
}

File: lib/rust/src/tree.rs
//! From a chosen derivation to the result's tree and its warnings (engine
//! §12), and to the next stage's tokens (§11). Every walk here is
//! iterative.

use std::collections::{BTreeMap, BTreeSet};

use crate::earley::{Cap, EngineError, Frame, Recognizer, Tok};
use crate::fxhash::FxSet;
use crate::lower::{LEmit, LEmitItem, LTerm, Lowered};
use crate::rank::{DNode, Ranker};
use crate::result::{Node, NodeKind, Warning};
use crate::tags::{phoneme_of, SetId};

#[derive(Debug, Clone)]
pub(crate) enum IKind {
    Read { tok: u32, terminal: u32 },
    Close { prod: u32, start: u32, end: u32, tags: SetId, caps: Vec<Cap> },
}

#[derive(Debug, Clone)]
pub(crate) struct INode {
    pub kind: IKind,
    pub children: Vec<u32>,
}

/// A derivation as a plain tree: node 0 is the root, and every node's
/// children come after it.
pub(crate) struct ITree {
    pub nodes: Vec<INode>,
}

/// Builds the tree of a derivation.
pub(crate) fn build(ranker: &Ranker, root: u32) -> ITree {
    let mut nodes: Vec<INode> = Vec::new();
    let mut stack: Vec<(u32, Option<u32>)> = vec![(root, None)];
    while let Some((id, parent)) = stack.pop() {
        let index = nodes.len() as u32;
        match ranker.dag.arena[id as usize] {
            DNode::Read { tok, terminal, .. } => {
                nodes.push(INode { kind: IKind::Read { tok, terminal }, children: Vec::new() });
            }
            DNode::Close { body, set, item } => {
                let it = ranker.item(set, item);
                let tags = ranker.chart().sets[set as usize].tagset[item as usize];
                nodes.push(INode {
                    kind: IKind::Close {
                        prod: it.prod,
                        start: it.origin,
                        end: set,
                        tags,
                        caps: ranker.chart().caps(it.caps).to_vec(),
                    },
                    children: Vec::new(),
                });
                let mut kids = Vec::new();
                let mut current = body;
                while let DNode::Seq { left, right } = ranker.dag.arena[current as usize] {
                    kids.push(right);
                    current = left;
                }
                // `kids` is last child first; the stack pops the first first.
                for kid in kids {
                    stack.push((kid, Some(index)));
                }
            }
            DNode::Seq { .. } | DNode::Empty => unreachable!("a derivation's nodes are reads and closes"),
        }
        if let Some(parent) = parent {
            nodes[parent as usize].children.push(index);
        }
    }
    ITree { nodes }
}

/// What the public tree needs to know of a stage: its tokens and tag names.
pub(crate) struct TreeContext<'a> {
    pub g: &'a Lowered,
    pub tokens: &'a [Tok],
    pub tag_map: &'a dyn Fn(SetId) -> BTreeMap<String, bool>,
    /// For the `elision-only` check: which tokens are written-back
    /// terminators, to be shown as elided nodes over the original input.
    pub synthetic: Option<&'a [bool]>,
}

impl<'a> TreeContext<'a> {
    fn original(&self, index: u32) -> usize {
        match self.synthetic {
            None => index as usize,
            Some(flags) => index as usize - flags[..index as usize].iter().filter(|&&flag| flag).count(),
        }
    }

    fn original_tokens(&self) -> Vec<&'a Tok> {
        match self.synthetic {
            None => self.tokens.iter().collect(),
            Some(flags) => self.tokens.iter().zip(flags).filter(|(_, &flag)| !flag).map(|(token, _)| token).collect(),
        }
    }
}

/// The source range of a span of tokens: from the first's start to the
/// last's end, or, for an empty span, empty at the end of the token before.
pub(crate) fn source_of(tokens: &[&Tok], start: usize, end: usize) -> std::ops::Range<usize> {
    if start < end {
        tokens[start].source.0..tokens[end - 1].source.1
    } else {
        let at = if start > 0 { tokens[start - 1].source.1 } else { tokens.first().map_or(0, |token| token.source.0) };
        at..at
    }
}

/// Builds the result's tree (§12): helpers and the prefixes of trailing
/// repetitions spliced out, absent elidable optionals as elided nodes.
pub(crate) fn public_tree(tree: &ITree, context: &TreeContext) -> Node {
    let originals = context.original_tokens();
    let mut fragments: Vec<Vec<Node>> = (0..tree.nodes.len()).map(|_| Vec::new()).collect();
    for index in (0..tree.nodes.len()).rev() {
        let node = &tree.nodes[index];
        let made = match node.kind {
            IKind::Read { tok, terminal } => {
                let at = context.original(tok);
                let name = context.g.terminals[terminal as usize].clone();
                let synthetic = context.synthetic.is_some_and(|flags| flags[tok as usize]);
                let source = source_of(&originals, at, if synthetic { at } else { at + 1 });
                vec![Node {
                    kind: if synthetic { NodeKind::Elided } else { NodeKind::Token },
                    rule: None,
                    terminal: Some(name),
                    token: if synthetic { None } else { Some(at) },
                    span: at..if synthetic { at } else { at + 1 },
                    source,
                    tags: BTreeMap::new(),
                    children: Vec::new(),
                }]
            }
            IKind::Close { prod, start, end, tags, .. } => {
                let production = &context.g.prods[prod as usize];
                let rule = &context.g.rules[production.rule as usize];
                let mut children = Vec::new();
                for (position, &child) in node.children.iter().enumerate() {
                    let mut made = std::mem::take(&mut fragments[child as usize]);
                    if position == 0 && production.trailing_step && made.len() == 1 && made[0].kind == NodeKind::Rule {
                        // Take over the prefix's list rather than copy it, so
                        // that a long repetition costs linear time.
                        children = std::mem::take(&mut made[0].children);
                    } else {
                        children.append(&mut made);
                    }
                }
                let (start, end) = (context.original(start), context.original(end));
                if rule.helper {
                    match (&rule.elided, production.syms.is_empty()) {
                        (Some(terminal), true) => vec![Node {
                            kind: NodeKind::Elided,
                            rule: None,
                            terminal: Some(terminal.clone()),
                            token: None,
                            span: start..start,
                            source: source_of(&originals, start, start),
                            tags: BTreeMap::new(),
                            children: Vec::new(),
                        }],
                        _ => children,
                    }
                } else {
                    vec![Node {
                        kind: NodeKind::Rule,
                        rule: Some(rule.name.clone()),
                        terminal: None,
                        token: None,
                        span: start..end,
                        source: source_of(&originals, start, end),
                        tags: (context.tag_map)(tags),
                        children,
                    }]
                }
            }
        };
        fragments[index] = made;
    }
    let mut root = std::mem::take(&mut fragments[0]);
    root.pop().expect("the root is a rule node")
}

/// The warnings of a stage's chosen tree (§12): each rule node gives one
/// for each warning of its production whose feature is on, in the order a
/// walk meets the nodes, parent before children and children left to
/// right. The walk passes through what the tree splices out, helpers and
/// the prefixes of trailing repetitions, without counting them as nodes,
/// and so meets the tree's nodes in the tree's own order.
pub(crate) fn warnings_of(
    tree: &ITree,
    g: &Lowered,
    tokens: &[Tok],
    features: &BTreeSet<String>,
    stage: &str,
) -> Vec<Warning> {
    let tokens: Vec<&Tok> = tokens.iter().collect();
    let mut warnings = Vec::new();
    // Each node with whether the tree splices it out as a prefix.
    let mut stack = vec![(0u32, false)];
    while let Some((index, prefix)) = stack.pop() {
        let node = &tree.nodes[index as usize];
        let IKind::Close { prod, start, end, .. } = node.kind else {
            continue;
        };
        let production = &g.prods[prod as usize];
        if !prefix {
            // A helper's productions have no warnings.
            for feature in production.warnings.iter().filter(|&feature| features.contains(feature)) {
                warnings.push(Warning {
                    stage: stage.to_string(),
                    feature: feature.clone(),
                    rule: g.rules[production.rule as usize].name.clone(),
                    span: start as usize..end as usize,
                    source: source_of(&tokens, start as usize, end as usize),
                });
            }
        }
        for (position, &child) in node.children.iter().enumerate().rev() {
            stack.push((child, position == 0 && production.trailing_step));
        }
    }
    warnings
}

/// A token emitted for the next stage.
#[derive(Debug, Clone)]
pub(crate) struct Emitted {
    pub span: (usize, usize),
    pub source: (usize, usize),
    pub tags: SetId,
    pub phonemes: Option<String>,
    /// Whether its phonemes are its text (§11).
    pub verbatim: bool,
    pub inserted_by: Option<String>,
}

enum Work {
    Visit(u32),
    /// A token covering node `node` with the given tags.
    Cover(u32, SetId),
    /// An inserted token with one tag, at a token index, in a node.
    Insert {
        tag: String,
        at: u32,
        node: u32,
    },
}

fn span_of(tree: &ITree, index: u32) -> (u32, u32) {
    match tree.nodes[index as usize].kind {
        IKind::Read { tok, .. } => (tok, tok + 1),
        IKind::Close { start, end, .. } => (start, end),
    }
}

/// The phoneme of a token's strong phoneme tag, if it has one (§5). Two
/// such tags are an error of the grammar on any token, verbatim or not.
fn phoneme_tag(recognizer: &Recognizer, tags: SetId) -> Result<Option<String>, EngineError> {
    let mut own: Option<&str> = None;
    for &(id, strong) in recognizer.shared.tags.list(tags) {
        if let (true, Some(phoneme)) = (strong, phoneme_of(recognizer.shared.tags.name(id))) {
            if own.is_some() {
                return Err(EngineError { message: "a token carries two strong phoneme tags".to_string(), rule: None });
            }
            own = Some(phoneme);
        }
    }
    Ok(own.map(str::to_string))
}

/// What a node says (§5): the phonemes of the tokens below it, skipping
/// every constituent that emits ε, the node's own included. It keeps only
/// the first of each run of pause tokens, and leaves out a pause token at
/// either end. It counts pauses by token, so a verbatim token keeps its
/// periods.
fn spoken(g: &Lowered, tree: &ITree, tokens: &[Tok], root: u32) -> String {
    let mut pieces: Vec<&str> = Vec::new();
    let mut stack = vec![root];
    while let Some(index) = stack.pop() {
        let node = &tree.nodes[index as usize];
        match node.kind {
            IKind::Read { tok, .. } => {
                let phonemes = tokens[tok as usize].phonemes.as_deref().unwrap_or("");
                let repeated = phonemes == "." && matches!(pieces.last(), None | Some(&"."));
                if !phonemes.is_empty() && !repeated {
                    pieces.push(phonemes);
                }
            }
            IKind::Close { prod, .. } => {
                // A constituent that emits ε does not count (§11).
                if !matches!(g.prods[prod as usize].emit, LEmit::Nothing) {
                    stack.extend(node.children.iter().rev());
                }
            }
        }
    }
    if pieces.last() == Some(&".") {
        pieces.pop();
    }
    pieces.concat()
}

/// The source position of an empty span at the token index `at`: the
/// source end of the token before it, or the source start of the first.
fn empty_source(tokens: &[Tok], at: u32) -> usize {
    if at > 0 {
        tokens[at as usize - 1].source.1
    } else {
        tokens.first().map_or(0, |token| token.source.0)
    }
}

/// The token that covers node `index` with the given tags (§5, §11).
/// `widened_ends` holds where the widened tokens emitted before it end.
fn cover(
    recognizer: &Recognizer,
    tree: &ITree,
    tokens: &[Tok],
    index: u32,
    tags: SetId,
    widened_ends: &mut FxSet<u32>,
) -> Result<Emitted, EngineError> {
    let (start, end) = span_of(tree, index);
    let span = (start as usize, end as usize);
    // Two phoneme tags are an error on any token, verbatim or not (§5).
    let phoneme = phoneme_tag(recognizer, tags)?;
    let text = recognizer.shared.text;
    if let IKind::Close { prod, .. } = tree.nodes[index as usize].kind {
        if recognizer.g.prods[prod as usize].verbatim {
            // A widened token takes in the text next to it that no input
            // token covers, but not text that a widened token before it
            // took. It sounds like its text.
            let before = if start > 0 { tokens[span.0 - 1].source.1 } else { 0 };
            let source = if start == end {
                (before, before)
            } else {
                // It always holds its own tokens' sources, which the tokens
                // next to it can share.
                let own = (tokens[span.0].source.0, tokens[span.1 - 1].source.1);
                let from = if widened_ends.contains(&start) { own.0 } else { own.0.min(before) };
                let after = tokens.get(span.1).map_or(text.len(), |token| token.source.0);
                widened_ends.insert(end);
                (from, own.1.max(after))
            };
            let phonemes = text[source.0..source.1].iter().collect();
            return Ok(Emitted { span, source, tags, phonemes: Some(phonemes), verbatim: true, inserted_by: None });
        }
    }
    if end - start == 1 && tokens[span.0].verbatim {
        // A token over one verbatim token is verbatim, with its source.
        let only = &tokens[span.0];
        let phonemes = Some(only.text.clone());
        return Ok(Emitted { span, source: only.source, tags, phonemes, verbatim: true, inserted_by: None });
    }
    let source = if start < end {
        (tokens[span.0].source.0, tokens[span.1 - 1].source.1)
    } else {
        let at = empty_source(tokens, start);
        (at, at)
    };
    let phonemes = phoneme.unwrap_or_else(|| spoken(recognizer.g, tree, tokens, index));
    Ok(Emitted { span, source, tags, phonemes: Some(phonemes), verbatim: false, inserted_by: None })
}

/// The tags an emission item's term gives a token (§11): a term that gives
/// none is an error of the grammar, since no terminal could read the token.
fn item_tags(recognizer: &mut Recognizer, term: &LTerm, frame: &Frame, tokens: &[Tok]) -> Result<SetId, EngineError> {
    let set = recognizer.tag_term(term, frame, tokens, 0)?;
    if recognizer.shared.tags.list(set).is_empty() {
        let owner = recognizer.g.prods[frame.prod as usize].owner;
        return Err(EngineError {
            message: "an emission gives a token no tags, which no terminal can read; %emits ε emits nothing"
                .to_string(),
            rule: Some(owner),
        });
    }
    Ok(set)
}

/// Walks the chosen tree from the left and emits the next stage's tokens
/// (§11).
pub(crate) fn emit(recognizer: &mut Recognizer, tree: &ITree, tokens: &[Tok]) -> Result<Vec<Emitted>, EngineError> {
    let g = recognizer.g;
    let mut out = Vec::new();
    // Where the widened tokens emitted so far end (§11).
    let mut widened_ends: FxSet<u32> = FxSet::default();
    let mut stack = vec![Work::Visit(0)];
    while let Some(work) = stack.pop() {
        match work {
            Work::Visit(index) => {
                let node = &tree.nodes[index as usize];
                let IKind::Close { prod, start, end, tags, caps } = &node.kind else {
                    continue;
                };
                let (tags, caps) = (*tags, caps.clone());
                let frame = Frame { caps: &caps, prod: *prod, origin: *start, end: *end, tags: Some(tags) };
                let production = &g.prods[*prod as usize];
                match &production.emit {
                    LEmit::Nothing => {}
                    LEmit::None => {
                        for &child in node.children.iter().rev() {
                            stack.push(Work::Visit(child));
                        }
                    }
                    LEmit::This(items) => {
                        let mut covers = Vec::new();
                        for term in items {
                            let set = match term {
                                Some(term) => item_tags(recognizer, term, &frame, tokens)?,
                                None => tags,
                            };
                            covers.push(Work::Cover(index, set));
                        }
                        stack.extend(covers.into_iter().rev());
                    }
                    LEmit::Items(items) => {
                        // Exactly the items, in the order listed; nothing
                        // inside the constituent is walked (§11).
                        let (_, end) = span_of(tree, index);
                        let child = |slot: u8| node.children[production.cap_pos[slot as usize] as usize];
                        let mut sequence = Vec::with_capacity(items.len());
                        for item in items {
                            match item {
                                LEmitItem::Cap(slot, term) => {
                                    let set = match term {
                                        Some(term) => item_tags(recognizer, term, &frame, tokens)?,
                                        None => caps[*slot as usize].tags,
                                    };
                                    sequence.push(Work::Cover(child(*slot), set));
                                }
                                LEmitItem::Insert(tag, anchor) => {
                                    // At the start of the part of the capture
                                    // listed next, or at the constituent's end.
                                    let at = anchor.map_or(end, |slot| span_of(tree, child(slot)).0);
                                    sequence.push(Work::Insert { tag: tag.clone(), at, node: index });
                                }
                            }
                        }
                        stack.extend(sequence.into_iter().rev());
                    }
                }
            }
            Work::Cover(index, tags) => {
                out.push(cover(recognizer, tree, tokens, index, tags, &mut widened_ends)?);
            }
            Work::Insert { tag, at, node } => {
                let (start, end) = span_of(tree, node);
                let source = if at > start {
                    tokens[at as usize - 1].source.1
                } else if start < end {
                    tokens[start as usize].source.0
                } else {
                    empty_source(tokens, start)
                };
                let IKind::Close { prod, .. } = &tree.nodes[node as usize].kind else { unreachable!("a close") };
                let owner = g.prods[*prod as usize].owner;
                let set = recognizer.shared.tags.set_of([(tag.as_str(), true)]);
                out.push(Emitted {
                    span: (at as usize, at as usize),
                    source: (source, source),
                    tags: set,
                    phonemes: Some(phoneme_of(&tag).unwrap_or("").to_string()),
                    verbatim: false,
                    inserted_by: Some(g.rules[owner as usize].name.clone()),
                });
            }
        }
    }
    Ok(out)
}
